### PR TITLE
feat: `--mode contract-call` added

### PR DESCRIPTION
**Pull request by IdrisHanafi**
# Description

This PR introduces a new polycli loadtest mode, `--mode contract-call`, which allows the calling of any arbitrary contract method. In order to use `--mode contract-call`, the following must be passed in:
- `--contract-address`
- `--calldata`
- if the contract function is `payable`, then the `--contract-call-payable` should also be passed in and it will use the `--eth-amount` flag as the value passed in.

Essentially, invoking `cast call` or `cast send` in a loadtest.

Some follow-up tasks would be to add:
- [ ] Allow function signatures and args to be passed like `--function-sig "setNumber(uint)" --args 10`

## Jira / Linear Tickets

- [DVT-1072](https://polygon.atlassian.net/browse/DVT-1072)

# Testing

## tl;dr

A simple non-state changing call (i.e. no transactions created):
```
$ cast calldata "getNumber()(uint)"
0xf2c9ecd8
$ go run main.go loadtest \
                --verbosity 700 \
                --mode contract-call \
                --contract-address $CONTRACT_ADDRESS \
                --calldata "0xf2c9ecd8" \
                --call-only \
                --rpc-url http://localhost:8545
```

Example state-changing (i.e. sends a transaction) function call with args:
```
$ cast calldata "setNumber(uint)" 21
0x3fb5c1cb0000000000000000000000000000000000000000000000000000000000000015
$ go run main.go loadtest \
                --verbosity 700 \
                --mode contract-call \
                --contract-address $CONTRACT_ADDRESS \
                --calldata "0x3fb5c1cb0000000000000000000000000000000000000000000000000000000000000015" \
                --rpc-url http://localhost:8545
```

A `payable` function:
```
$ cast calldata "storeEther()"
0xa7cc190a
$ go run main.go loadtest \
                --verbosity 700 \
                --mode cc \
                --contract-address $CONTRACT_ADDRESS \
                --contract-call-payable \
                --calldata "0xa7cc190a" \
                --eth-amount 0.00000003 \
                --rpc-url http://localhost:8545
```

A multiarg call:
```
$ cast calldata "getMultipliedAndAddNumber(uint,uint)(uint)" 10 3
0xdd0974e2000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000003
go run main.go loadtest \
                --verbosity 700 \
                --mode contract-call \
                --contract-address $CONTRACT_ADDRESS \
                --calldata "0xdd0974e2000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000003" \
                --call-only \
                --rpc-url http://localhost:8545
```

## Test Process

Testing this, I had the following in mind:
1. Create a simple Solidity contract that has functions that are of mixtures of a free method, non-free method, payable, and multi-arg.
2. Generate `bin`
3. Deploy the sample contract with `cast send`
4. Test calls for the different types of methods in `polycli loadtest --mode contract-call`

Before we start, let's run `geth` in dev mode and fund our loadtest account `0x85da99c8a7c2c95964c8efd687e95e632fc533d6` (`code code...code code quality` mnemonic):
```
$ geth --dev --dev.period 2 --http --rpc.gascap 5000000000 --miner.gaslimit 5000000000 --dev.gaslimit 5000000000 --http.api 'admin,debug,web3,eth,txpool,personal,miner,net
$ geth attach "http://127.0.0.1:8545"
> eth.sendTransaction({from: eth.coinbase, to: "0x85da99c8a7c2c95964c8efd687e95e632fc533d6", value: web3.toWei(50000000000, "ether")})
```

### 1. Create Solidity Contract

This is a sample contract that I used:
```
// SPDX-License-Identifier: GPL-3.0
pragma solidity ^0.8.23;

contract MinimalContractTest {
    uint public myNumber;
    mapping(address => uint) public balances;

    function storeEther() public payable {
        require(msg.value > 0, "Must send more than 0 ether");
        balances[msg.sender] += msg.value;
    }

    function checkBalance() public view returns (uint) {
        return balances[msg.sender];
    }

    function withdrawEther() public {
        uint balance = balances[msg.sender];
        require(balance <= 0, "No Ether to withdraw");

        balances[msg.sender] = 0;

        (bool sent, ) = msg.sender.call{value: balance}("");
        require(sent, "Failed to send Ether");
    }

    function getContractBalance() public view returns (uint) {
        return address(this).balance;
    }

    function setNumber(uint _number) public {
        myNumber = _number;
    }

    function getNumber() public view returns (uint) {
        return myNumber;
    }

    function getMultipliedNumber(uint multiplier) public view returns (uint) {
        return multiplier * myNumber;
    }

    function getMultipliedAndAddNumber(uint multiplier, uint adder) public view returns (uint) {
        return myNumber * multiplier + adder;
    }
}
```

### 3. Generate `bin`

Using `forge`:
```
FOUNDRY_PROFILE=lite forge build --force
```

### 3. Deploy the contract with `cast send`
```
CONTRACT_ADDRESS=`cast send \
    --from 0x85dA99c8a7C2C95964c8EfD687E95E632Fc533D6 \
    --private-key 0x42b6e34dc21598a807dc19d7784c71b2a7a01f6480dc6f58258f78e539f1a1fa \
    --rpc-url localhost:8545 \
    --json \
    --create \
    "$(cat out/MinimalContractTest.sol/MinimalContractTest.json | jq -r '.bytecode.object')" | jq -r ".contractAddress"`
echo $CONTRACT_ADDRESS
```

### 4. Test calls for the different types of methods in `polycli loadtest --mode contract-call`

A simple non-state changing call (i.e. no transactions created):
```
$ cast calldata "getNumber()(uint)"
0xf2c9ecd8
$ go run main.go loadtest \
                --verbosity 700 \
                --mode contract-call \
                --contract-address $CONTRACT_ADDRESS \
                --calldata "0xf2c9ecd8" \
                --call-only \
                --rpc-url http://localhost:8545
```

Example state-changing (i.e. sends a transaction) function call with args:
```
$ cast calldata "setNumber(uint)" 21
0x3fb5c1cb0000000000000000000000000000000000000000000000000000000000000015
$ go run main.go loadtest \
                --verbosity 700 \
                --mode contract-call \
                --contract-address $CONTRACT_ADDRESS \
                --calldata "0x3fb5c1cb0000000000000000000000000000000000000000000000000000000000000015" \
                --rpc-url http://localhost:8545
```

A `payable` function:
```
$ cast calldata "storeEther()"
0xa7cc190a
$ go run main.go loadtest \
                --verbosity 700 \
                --mode cc \
                --contract-address $CONTRACT_ADDRESS \
                --contract-call-payable \
                --calldata "0xa7cc190a" \
                --eth-amount 0.00000003 \
                --rpc-url http://localhost:8545
```

A multiarg call:
```
$ cast calldata "getMultipliedAndAddNumber(uint,uint)(uint)" 10 3
0xdd0974e2000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000003
go run main.go loadtest \
                --verbosity 700 \
                --mode contract-call \
                --contract-address $CONTRACT_ADDRESS \
                --calldata "0xdd0974e2000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000003" \
                --call-only \
                --rpc-url http://localhost:8545
```

To validate it's actually working, check initial balance:
```
$ cast call $CONTRACT_ADDRESS "checkBalance()(uint)" --mnemonic 'code code code code code code code code code code code quality'
0
```

Then store some ether using the loadtest:
```
$ cast calldata "storeEther()"
0xa7cc190a
$ go run main.go loadtest \
                --verbosity 700 \
                --mode cc \
                --contract-address $CONTRACT_ADDRESS \
                --contract-call-payable \
                --calldata "0xa7cc190a" \
                --eth-amount 0.00000003 \
                --rate-limit 5 \
                --concurrency 5 \
                --requests 5 \
                --rpc-url http://localhost:8545
```
> NOTE: `storeEther()` is payable so `--contract-call-payable` is passed

Then check balances again:
```
$ cast call $CONTRACT_ADDRESS "checkBalance()(uint)" --mnemonic 'code code code code code code code code code code code quality'
749999999975 [7.499e11]
```

- [x] non-state changing call (i.e. no transactions created) working
- [x] state-changing (i.e. sends a transaction) function call with args working
- [x] `payable` function call working
- [x] multiarg function call working

[DVT-1072]: https://polygon.atlassian.net/browse/DVT-1072?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: cmd/loadtest/app.go [hunk 1]
```diff
@@ -237,14 +241,18 @@ r - random modes
...
 	ltp.ByteCount = LoadtestCmd.Flags().Uint64P("byte-count", "b", 1024, "If we're in store mode, this controls how many bytes we'll try to store in our contract")
 	ltp.LtAddress = LoadtestCmd.Flags().String("lt-address", "", "The address of a pre-deployed load test contract")
 	ltp.ERC20Address = LoadtestCmd.Flags().String("erc20-address", "", "The address of a pre-deployed ERC20 contract")
 	ltp.ERC721Address = LoadtestCmd.Flags().String("erc721-address", "", "The address of a pre-deployed ERC721 contract")
 	ltp.ForceContractDeploy = LoadtestCmd.Flags().Bool("force-contract-deploy", false, "Some load test modes don't require a contract deployment. Set this flag to true to force contract deployments. This will still respect the --lt-address flags.")
 	ltp.RecallLength = LoadtestCmd.Flags().Uint64("recall-blocks", 50, "The number of blocks that we'll attempt to fetch for recall")
+	ltp.ContractAddress = LoadtestCmd.Flags().String("contract-address", "", "The address of the contract that will be used in `--mode contract-call`. This must be paired up with `--mode contract-call` and `--calldata`")
```
Comment (leovct) [inline cmd/loadtest/app.go:253]: I just discovered cobra's `MarkFlag` and `MarkFlags` methods and I think it's very handy, especially in such cases. 

Screenshot 2023-12-05 at 09 23 47

You can mark those 3 flags required together using [`MarkFlagsRequiredTogether`](https://pkg.go.dev/github.com/spf13/cobra#Command.MarkFlagsRequiredTogether). If the user specifies one or two of the three flag but not all three, it would throw an error.

File: cmd/loadtest/loadtest.go [hunk 2]
```diff
@@ -236,6 +243,7 @@ func initializeLoadTestParams(ctx context.Context, c *ethclient.Client) error {
 	}
 
 	if len(modes) > 1 {
+		log.Debug().Msgf("FOUND MORE THAN 1 MODE %+v", modes)
```
Comment (leovct) [inline cmd/loadtest/loadtest.go:246]: Should we clean this up?

File: cmd/loadtest/loadtest.go [hunk 3]
```diff
@@ -295,6 +306,7 @@ func completeLoadTest(ctx context.Context, c *ethclient.Client, rpc *ethrpc.Clie
 		log.Error().Err(err).Msg("There was an issue waiting for all transactions to be mined")
 	}
 	if len(loadTestResults) == 0 {
+		log.Info().Msg("CallOnly mode enabled - blocks aren't mined")
```
Comment (leovct) [inline cmd/loadtest/loadtest.go:309]: I don't think that's true all the time. Sometimes transactions are not observed because of a problem with the RPC or something else? It would maybe worth adding another if branch that checks if the user specified `call-only` and return `nil` here?

**Comment by leovct** [comment]
Also just a nit, I learnt this while working with UniswapV3 and the PoS bridge contracts :)
When using `cast send`, you don't need to specify the `--from` flag since it will be derived from `--private-key`.

File: cmd/loadtest/loadtest.go [hunk 4]
```diff
@@ -1233,6 +1247,84 @@ func loadTestRPC(ctx context.Context, c *ethclient.Client, nonce uint64, ia *Ind
...
+	privateKey := ltp.ECDSAPrivateKey
+
+	tops, err := bind.NewKeyedTransactorWithChainID(privateKey, chainID)
+	if err != nil {
+		log.Error().Err(err).Msg("Unable create transaction signer")
+		return
+	}
+
+	amount := big.NewInt(0)
+	if *ltp.ContractCallPayable {
```
Comment (leovct) [inline cmd/loadtest/loadtest.go:1265]: I was just testing and thought that maybe we could remove this flag? Let's say the user specified `--mode contract-call --contract-address <...> --calldata <...>` and also `--eth-amount`, then you know you're going to send a transaction with some eth so maybe `--contract-call-payable` is a bit redundant here?

**Comment by leovct** [comment]
I wanted to send a little more eth from the loadtest account to the contract and it reverts. That's weird because I have enough ether... I don't think that's related to your feature though, I can investigate later :)

Screenshot 2023-12-05 at 10 15 46

**Review by leovct — APPROVED** [review]

File: cmd/loadtest/loadtest.go (hunk 3 above)
Comment (IdrisHanafi) [reply cmd/loadtest/loadtest.go:309]: Woah, good catch. I have 100% forgot to add that if-statement 😂 

Updated. :)

File: cmd/loadtest/app.go (hunk 1 above)
Comment (IdrisHanafi) [reply cmd/loadtest/app.go:253]: neat, good find! I'll try to utilize this soon :)

**Comment by IdrisHanafi** [comment]
> I wanted to send a little more eth from the loadtest account to the contract and it reverts. That's weird because I have enough ether... I don't think that's related to your feature though, I can investigate later :)
> 
> Screenshot 2023-12-05 at 10 15 46

Yeah that's odd....I'll investigate into this later too.

File: cmd/loadtest/loadtest.go (hunk 4 above)
Comment (leovct) [reply cmd/loadtest/loadtest.go:1267]: Okay makes sense :)

**Comment by IdrisHanafi** [comment]
@leovct the bug was that for estimating gas, I had to specify the `From` address within the `CallMsg`. Pushed the updated and reran against the `payable` function and it works.

**Review by leovct — APPROVED** [review]
lgtm!!
